### PR TITLE
🐛 dy-sidecar sometimes does not save state and outputs

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- ♻️ changed some function naming to make it more easy to establish what they do without opening them
- 🐛 If an error is triggered during the observation-cycle, the `can_save` property was wrongly computed and would cause data loss. This is what happened in the issue described here https://github.com/ITISFoundation/osparc-simcore/issues/3413

**NOTE:** This issue is present since a very long time, but was never reported before.

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3413
- https://github.com/ITISFoundation/osparc-issues/issues/638

### CODE REVIEW

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task_utils.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ async def apply_observation_cycle(
         )
         await scheduler.mark_service_for_removal(
             node_uuid=scheduler_data.node_uuid,
-            can_save=scheduler_data.dynamic_sidecar.service_removal_state.can_save,
+            can_save=scheduler_data.dynamic_sidecar.were_containers_created,
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task_utils.py:42]: so this is the the fix in here?

**Review by pcrespov — APPROVED** [review]
I hope this fixes it :-) 

I still have my concerns regarding the reliability of the states logic on the directorv2 service and the IMO unconventional naming scheme that you use in there. I guess at some point we should catch up on those points again and share thoughts ... in the meantime, go ahead :-)

**Review by elisabettai — COMMENTED** [review]
:+1:

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/__init__.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 from ._core import (
-    are_all_services_present,
+    all_dynamic_sidecar_stack_services_present,
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/__init__.py:2]: ;) not sure I get better what is exactly meant here...
all_services_present
vs
all_dynamic_sidecar_stack_services_present

Without looking in the code, now I'm wondering you mean the actual service (as seen in the osparc GUI) or the services + sidecar + proxy?
I guess it is the latter right? you could define it as a ```pod``` ?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py [hunk 3]
```diff
@@ -89,13 +89,15 @@ def parse_containers_inspect(
...
+def are_all_service_containers_running(
+    containers_inspect: List[DockerContainerInspect],
+) -> bool:
     return len(containers_inspect) > 0 and all(
         (x.status == DockerStatus.RUNNING for x in containers_inspect)
     )
 
 
-async def attempt_user_create_services_removal_and_data_saving(
+async def attempt_user_created_services_removal_and_data_saving(
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py:100]: for the naming above, then you could use a convention like:
- user_services = the car or the service as in the GUI
- sidecar, proxy = obvious
- pod = sidecar,proxy+user_services

**Review by sanderegg — APPROVED** [review]
I think it could maybe beneficial to define specific names for this sidecar+proxy+user_containers/services architecture...

Great, and thanks for the fix!

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task_utils.py (hunk 1 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task_utils.py:42]: Yes, exactly there. This is a long time bug.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py (hunk 3 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py]: > * user_services = the car or the service as in the GUI
> * sidecar, proxy = obvious

I like these a lot

> * pod = sidecar,proxy+user_services

I like it less but it does the job

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/__init__.py (hunk 2 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/__init__.py]: Below convention was adapted and function was renamed

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py (hunk 3 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_utils.py]: Let's go with this, I've put it int he `__init__.py` of the `dynamic_sidecar` module of `director-v2`

**Pull request merged**